src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,243 @@
+                                var parameters = methodSymbol.GetParameters();
+                                var flagHasDeserializeAttributes = attributeTypeSymbols.Length != 0
+                                    && attributeTypeSymbols.Any(s => methodSymbol.HasAttribute(s))
+                                    && parameters.Length == 1
+                                    && parameters[0].Type.Equals(streamingContextTypeSymbol);
+                                var flagImplementOnDeserializationMethod = methodSymbol.IsOnDeserializationImplementation(IDeserializationCallback);
+                                var flagImplementDisposeMethod = methodSymbol.IsDisposeImplementation(compilation);
+                                var flagIsFinalizer = methodSymbol.IsFinalizer();
+
+                                if (!flagHasDeserializeAttributes && !flagImplementOnDeserializationMethod && !flagImplementDisposeMethod && !flagIsFinalizer)
+                                {
+                                    continue;
+                                }
+
+                                foreach (var result in results[methodSymbol])
+                                {
+                                    compilationAnalysisContext.ReportDiagnostic(
+                                        methodSymbol.CreateDiagnostic(
+                                            Rule,
+                                            methodSymbol.ContainingType.MetadataName,
+                                            methodSymbol.MetadataName,
+                                            result.MetadataName));
+                                }
+                            }
+                        });
+
+                    /// <summary>
+                    /// Analyze the method to find all the dangerous method it calls.
+                    /// </summary>
+                    /// <param name="methodSymbol">The symbol of the method to be analyzed</param>
+                    /// <param name="visited">All the method has been analyzed</param>
+                    /// <param name="results">The result is organized by &lt;method to be analyzed, dangerous method it calls&gt;</param>
+                    void FindCalledDangerousMethod(IMethodSymbol methodSymbol, HashSet<IMethodSymbol> visited, Dictionary<IMethodSymbol, HashSet<IMethodSymbol>> results)
+                    {
+                        if (visited.Add(methodSymbol))
+                        {
+                            results.Add(methodSymbol, new HashSet<IMethodSymbol>());
+
+                            foreach (var child in callGraph[methodSymbol].Keys)
+                            {
+                                if (dangerousMethodSymbols.Contains(child))
+                                {
+                                    results[methodSymbol].Add(child);
+                                }
+
+                                if (child.IsInSource())
```

Comment:
For people compiling .NET base class libraries, should be: `else if (child.IsInSource())`.